text/0001-blinking-cursors-and-periodic-animation.md

```diff
@@ -0,0 +1,182 @@
+2. Indicate to the runtime that an update is needed in 500ms increments after the last user interaction
+3. Conditionally emit display list commands to blink the cursor: omitting cursor primitives during the 'hide' phase of the animation (1/2 second), and including cursor primitives during the 'show' phase of the animation (other 1/2 second).
+
+# Reference-level explanation
+
+### TL;DR
+
+ * The problem of a blinking cursor input can be generalized to _periodic animation_ support.
+ * The `iced` runtime only performs an update/draw cycle in response to application events and user interaction, but to support periodic animation, we need the runtime to perform a cycle whenever a widget needs it.
+ * The current interface (the [Widget](https://docs.rs/iced_native/0.2.2/iced_native/widget/trait.Widget.html) trait) between the runtime and widgets has no means to indicate that a widget is animating, and hence requires a re-draw at sometime in the future.
+
+_The numbering of these points aligns to that of the guided explanation in the previous section_.
+
+### 1: Event loop changes
+
+The event loop needs to be changed to track the soonest moment which an update needs to occur, and to perform an update/draw cycle at this moment.
+
+The intermediate state of a user interface (at the transition of a update/draw cycle) is stored in the `State` struct, from `native/src/program/state.rs`. A new field, `next_animation_draw: AnimationState` can be added to track the next draw required by widgets. More on the `AnimationState` type later, but just know that this type encapsulates the animation requirements of widgets, and this value can be obtained by calling a method on the root widget.
+
+Modifying the behavior of the wait state is surprisingly trivial, thanks to the underlying use of `winit`. Instead of setting the event loops' `control_flow` to `Wait`, we just set it to `WaitUntil( <time of soonest update> )`. This changes the behavior to wait until the next event, or the provided time (whichever is sooner).
+
+As the event loop only updates/draws in response to events/user-interaction, we emit a new synthetic event `AnimationTick` to drive the update cycle.
+
+### 2: Widget trait changes
+
+An update is needed to the [Widget](https://docs.rs/iced_native/0.2.2/iced_native/widget/trait.Widget.html) trait to communicate the animation requirements of a widget - in our case, that the `TextInput` widget needs to be re-drawn in 500ms.
+
+We propose adding a new method to the trait, with a default implementation that just indicate no animation is taking place:
+
+```rust
+  fn next_animation(&self) -> AnimationState {
+    AnimationState::NotAnimating
+  }
+```
+
+Widgets that need to animate (such as our `TextInput`) can implement this method to indicate an animation is required:
+
+```rust
+  fn next_animation(&self) -> AnimationState {
+    AnimationState::AnimateIn(std::time::Instant::now().checked_add(Duration::from_millis(500)))
```

Comment:
I think we should keep this method side-effect free. In other words, it should always return the same value if the implementor hasn't changed state. This becomes important when there are many widgets in the widget tree that want to animate at different times.

To achieve this, we could store the next animation instant in the `TextInput` state.

Reply:
Seems reasonable - the prototype ended up going with a 'almost side-effect free' approach:

https://github.com/twitchyliquid64/iced/commit/777efc85bf3cdbb6764cb8c8244ead4b5707b491#diff-ee7dbbf540eed4c5a7ca63c6dffb6eca511a4423edab4b5356ed243c777bb88dR546-R552

But we can totally make it an explicit value that is only updated actively. I'll make sure to update the RFC to represent this requirement.